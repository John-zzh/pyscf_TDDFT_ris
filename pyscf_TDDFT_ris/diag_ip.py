import numpy as np

def TDA_diag_initial_guess(V_holder, N_states, hdiag):
    '''
    N_states is the amount of initial guesses
    sort out the smallest value of hdiag, the corresponding position in the 
    initial guess set as 1.0, everywhere else set as 0.0
    '''
    hdiag = hdiag.reshape(-1,)
    Dsort = hdiag.argsort()
    for j in range(N_states):
        V_holder[Dsort[j], j] = 1.0
    return V_holder

def TDA_diag_preconditioner(residual, sub_eigenvalue, hdiag ):
    '''
    DX - XÎ© = r
    '''

    N_states = np.shape(residual)[1]
    t = 1e-14
    D = np.repeat(hdiag.reshape(-1,1), N_states, axis=1) - sub_eigenvalue
    '''
    force all small values not in [-t,t]
    '''
    D = np.where( abs(D) < t, np.sign(D)*t, D)
    new_guess = residual/D

    return new_guess

# def TDDFT_diag_initial_guess(V_holder, W_holder, N_states, hdiag):
#     '''
#     return unchanged N_states just to keep consistent with other funcs
#     '''
#     V_holder[:,:N_states] = TDA_diag_initial_guess(
#                                             V_holder = V_holder,
#                                             N_states = N_states,
#                                                hdiag = hdiag)

#     return V_holder, W_holder

def TDDFT_diag_preconditioner(R_x, R_y, omega, hdiag):
    '''
    preconditioners for each corresponding residual (state)
    '''
    hdiag = hdiag.reshape(-1,1)
    N_states = R_x.shape[1]
    t = 1e-14
    d = np.repeat(hdiag.reshape(-1,1), N_states, axis=1)

    D_x = d - omega
    D_x = np.where(abs(D_x) < t, np.sign(D_x)*t, D_x)
    D_x_inv = D_x**-1

    D_y = d + omega
    D_y = np.where(abs(D_y) < t, np.sign(D_y)*t, D_y)
    D_y_inv = D_y**-1

    X_new = R_x*D_x_inv
    Y_new = R_y*D_y_inv

    return X_new, Y_new

# def spolar_diag_initprec(RHS, hdiag=delta_hdiag2, conv_tol=None):
#
#     d = hdiag.reshape(-1,1)
#     RHS = RHS/d
#
#     return RHS
