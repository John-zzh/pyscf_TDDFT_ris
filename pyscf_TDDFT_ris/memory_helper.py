import os, psutil

def show_memory_info(hint):
    pid = os.getpid()
    p = psutil.Process(pid)
    info = p.memory_full_info()
    memory = info.uss / 1024**3
    print('{} memory used: {:<.2f} GB'.format(hint, memory))