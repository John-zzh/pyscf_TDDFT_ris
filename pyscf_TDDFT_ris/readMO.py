from pyscf import dft
import numpy as np

def get_mf_from_fch(fch_file, functional):
    '''
    fch_file: fch file generated by Gaussian
    functional: functional used in Gaussian
    '''
    from mokit.lib.gaussian import load_mol_from_fch, mo_fch2py
    from mokit.lib.rwwfn import read_eigenvalues_from_fch, read_nbf_and_nif_from_fch
    
    mol = load_mol_from_fch(fch_file)
    mo_coeff = mo_fch2py(fch_file)
    nbf, nif = read_nbf_and_nif_from_fch(fch_file)
    mo_energy = read_eigenvalues_from_fch(fch_file, nif=nif, ab='alpha')

    mol.build()
    print('======= Molecular Coordinates ========')
    print(mol.atom)

    # [print(k, v) for k, v in mol.basis.items()]
    mf = dft.RKS(mol)
    mf.mo_coeff = mo_coeff
    
    mf.mo_energy = mo_energy
    nocc = mol.nelectron // 2
    mf.mo_occ = np.asarray([2] * nocc + [0] * (nbf - nocc))
    mf.xc = functional
    return mf

def get_mf_from_molden(molden_file, functional, basis):
    '''
    molden_file: molden file generated by ORCA
    functional: functional used in ORCA
    basis: basis set name used in ORCA, 
    ORCA molden file does not contain unnoralized basis set information
    so we need to specify the basis set name and let pyscf to build the basis set
    '''
    from pyscf.tools import molden
    mol, mo_energy, mo_coeff, mo_occ, _, _ = molden.read(molden_file)

    mol.basis = basis
    mol.build()
    print('======= Molecular Coordinates ========')
    print(mol.atom)

    mf = dft.RKS(mol)
    mf.xc = functional
    mf.mo_occ = mo_occ
    mf.mo_coeff = mo_coeff
    mf.mo_energy = mo_energy

    return mf