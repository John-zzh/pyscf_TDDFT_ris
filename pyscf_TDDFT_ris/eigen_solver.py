
import numpy as np
from pyscf_TDDFT_ris import math_helper
from pyscf_TDDFT_ris import parameter
from scipy.sparse import csr_matrix
import time


def Davidson(matrix_vector_product,
                    hdiag,
                    N_states=20,
                    conv_tol=1e-5,
                    max_iter=25,
                    single=False ):
    '''
    AX = XΩ
    Davidson frame, can use different initial guess and preconditioner
    initial_guess is a function, takes the number of initial guess as input
    preconditioner is a function, takes the residual as the input
    '''
    print('====== Davidson Diagonalization Starts ======')

    D_start = time.time()

    A_size = hdiag.shape[0]
    print('size of A matrix =', A_size)
    size_old = 0
    size_new = min([N_states+8, 2*N_states, A_size])

    max_N_mv = max_iter*N_states + size_new
    V_holder = np.zeros((A_size, max_N_mv),dtype=np.float32 if single else np.float64)
    W_holder = np.zeros_like(V_holder)
    sub_A_holder = np.zeros((max_N_mv,max_N_mv),dtype=np.float32 if single else np.float64)
    '''
    generate the initial guesss and put into the basis holder V_holder
    '''
    V_holder = math_helper.TDA_diag_initial_guess(V_holder=V_holder, N_states=size_new, hdiag=hdiag)

    # V_holder[:,:size_new] = initial_vectors[:,:]

    subcost = 0
    MVcost = 0
    fill_holder_cost = 0
    subgencost = 0
    full_cost = 0
    print('step max||r||   sub_A.shape')
    for ii in range(max_iter):
    
        MV_start = time.time()
        W_holder[:, size_old:size_new] = matrix_vector_product(V_holder[:,size_old:size_new])
        MV_end = time.time()
        iMVcost = MV_end - MV_start
        MVcost += iMVcost

        subgencost_start = time.time()
        sub_A_holder = math_helper.gen_VW(sub_A_holder, V_holder, W_holder, size_old, size_new)
        sub_A = sub_A_holder[:size_new,:size_new]
        subgencost_end = time.time()
        subgencost += subgencost_end - subgencost_start
        sub_A = math_helper.utriangle_symmetrize(sub_A)
        # sub_A = math_helper.symmetrize(sub_A)

        '''
        Diagonalize the subspace Hamiltonian, and sorted.
        sub_eigenvalue[:N_states] are smallest N_states eigenvalues
        '''

        subcost_start = time.time()
        sub_eigenvalue, sub_eigenket = np.linalg.eigh(sub_A)
        sub_eigenvalue = sub_eigenvalue[:N_states]
        sub_eigenket = sub_eigenket[:,:N_states]
        subcost_end = time.time()
        subcost += subcost_end - subcost_start

        full_cost_start = time.time()
        full_guess = np.dot(V_holder[:,:size_new], sub_eigenket)
        full_cost_end = time.time()
        full_cost += full_cost_end - full_cost_start

        AV = np.dot(W_holder[:,:size_new], sub_eigenket)
        residual = AV - full_guess * sub_eigenvalue

        r_norms = np.linalg.norm(residual, axis=0).tolist()
        max_norm = np.max(r_norms)
        print('{:<3d}  {:<10.4e} {:<5d}'.format(ii+1, max_norm, sub_A.shape[0]))
        if max_norm < conv_tol or ii == (max_iter-1):
            break

        index = [r_norms.index(i) for i in r_norms if i>conv_tol]

        new_guess = math_helper.TDA_diag_preconditioner(residual = residual[:,index],
                                            sub_eigenvalue = sub_eigenvalue[index],
                                            hdiag = hdiag)

        fill_holder_cost_start = time.time()
        size_old = size_new
        V_holder, size_new = math_helper.Gram_Schmidt_fill_holder(V_holder, size_old, new_guess, double=True)
        fill_holder_cost_end = time.time()
        fill_holder_cost += fill_holder_cost_end - fill_holder_cost_start

    # energies = sub_eigenvalue*parameter.Hartree_to_eV

    D_end = time.time()
    Dcost = D_end - D_start

    if ii == max_iter-1:
        print('=== TDA Failed Due to Iteration Limit ===')
        print('current residual norms', r_norms)
        
    # print('energies:')
    # print(energies)
    print('Finished in {:d} steps, {:.2f} seconds'.format(ii+1, Dcost))
    print('Maximum residual norm = {:.2e}'.format(max_norm))
    print('Final subspace size = {:d}'.format(sub_A.shape[0]))
    for enrty in ['MVcost','fill_holder_cost','subgencost','subcost','full_cost']:
        cost = locals()[enrty]
        print("{:<10} {:<5.4f}s {:<5.2%}".format(enrty, cost, cost/Dcost))
    print('========== Davidson Diagonalization Done ==========')
    return sub_eigenvalue, full_guess

def Davidson_Casida(matrix_vector_product,
                        hdiag,
                        N_states=20,
                        conv_tol=1e-5,
                        max_iter=25,
                        GS=False,
                        single=False ):
    '''
    [ A B ] X - [1   0] Y Ω = 0
    [ B A ] Y   [0  -1] X   = 0

    '''
    print('======= TDDFT Eigen Solver Statrs =======')

    davidson_start = time.time()
    A_size = hdiag.shape[0]
    print('size of A matrix =', A_size)
    size_old = 0
    size_new = min([N_states+8, 2*N_states, A_size])

    max_N_mv = (max_iter+1)*N_states
    
    '''
    [U1] = [A B][V]
    [U2]   [B A][W]

    U1 = AV + BW
    U2 = AW + BV

    a = [V.T W.T][A B][V] = [V.T W.T][U1] = VU1 + WU2
                 [B A][W]            [U2]
    '''
    V_holder = np.zeros((max_N_mv, A_size),dtype=np.float32 if single else np.float64)
    W_holder = np.zeros_like(V_holder)

    U1_holder = np.zeros_like(V_holder)
    U2_holder = np.zeros_like(V_holder)

    VU1_holder = np.zeros((max_N_mv,max_N_mv),dtype=np.float32 if single else np.float64)
    VU2_holder = np.zeros_like(VU1_holder)
    WU1_holder = np.zeros_like(VU1_holder)
    WU2_holder = np.zeros_like(VU1_holder)

    VV_holder = np.zeros_like(VU1_holder)
    VW_holder = np.zeros_like(VU1_holder)
    WW_holder = np.zeros_like(VU1_holder)

    '''
    set up initial guess V= TDA initial guess, W=0
    '''

    V_holder = math_helper.TDA_diag_initial_guess(V_holder=V_holder,
                                                N_states=size_new,
                                                hdiag=hdiag)
    subcost = 0
    MVcost = 0
    fill_holder_cost = 0
    fill_holder_step_cost = 0
    subgencost = 0
    full_cost = 0
    # math_helper.show_memory_info('After Davidson initial guess set up')
    # print('step maximum residual norm')

    if GS == True:
        fill_holder = math_helper.VW_Gram_Schmidt_fill_holder
    else:
        fill_holder = math_helper.nKs_fill_holder

    for ii in range(max_iter):

        # print('size_old =', size_old)
        # print('size_new =', size_new)
        MV_start = time.time()

        U1_holder[size_old:size_new, :], U2_holder[size_old:size_new, :] = matrix_vector_product(
                                                                            X=V_holder[size_old:size_new, :],
                                                                            Y=W_holder[size_old:size_new, :])
        MV_end = time.time()
        MVcost += MV_end - MV_start

        subgenstart = time.time()



        '''
        generate the subspace matrices
        sub_A = np.dot
        '''
        (sub_A, sub_B, sigma, pi,
        VU1_holder, WU2_holder, VU2_holder, WU1_holder,
        VV_holder, WW_holder, VW_holder) = math_helper.gen_sub_ab(
                                                    V_holder, W_holder, U1_holder, U2_holder,
                                                    VU1_holder, WU2_holder, VU2_holder, WU1_holder,
                                                    VV_holder, WW_holder, VW_holder,
                                                    size_old, size_new)

        subgenend = time.time()
        subgencost += subgenend - subgenstart

        '''
        solve the eigenvalue omega in the subspace
        '''
        subcost_start = time.time()
        # pi = pi * -1
        omega, x, y = math_helper.TDDFT_subspace_eigen_solver(sub_A, sub_B, sigma, pi, N_states)
        # print('omega =', omega*parameter.Hartree_to_eV)
        subcost_end = time.time()
        subcost += subcost_end - subcost_start
        # print(sub_A, sub_B, sigma, pi)
        '''
        compute the residual
        R_x = U1x + U2y - X_full*omega
        R_y = U2x + U1y + Y_full*omega
        X_full = Vx + Wy
        Y_full = Wx + Vy
        '''
        full_cost_start = time.time()
 
        V = V_holder[:size_new,:]
        W = W_holder[:size_new,:]
        U1 = U1_holder[:size_new, :]
        U2 = U2_holder[:size_new, :]

        X_full = x.T @ V + y.T @ W
        Y_full = x.T @ W + y.T @ V

        R_x = x.T @ U1 + y.T @ U2 - omega.reshape(-1, 1) * X_full
        R_y = x.T @ U2 + y.T @ U1 + omega.reshape(-1, 1) * Y_full

        full_cost_end = time.time()
        full_cost += full_cost_end - full_cost_start

        residual = np.hstack((R_x, R_y))
        # print('residual size =', residual.shape)
        r_norms = np.linalg.norm(residual, axis=1).tolist()
        # print('r_norms.shape', len(r_norms))
        max_norm = np.max(r_norms)
        # print('{:<3d}  {:<10.4e}'.format(ii+1, max_norm))
        print(f'iter: {ii+1:<3d}  max|R|: {max_norm:<10.2e} new_vectors: {size_new - size_old}  MVP: {MV_end - MV_start:.1f} seconds')

        if max_norm < conv_tol or ii == (max_iter -1):
            # math_helper.show_memory_info('After last Davidson iteration')
            break

        index = [r_norms.index(i) for i in r_norms if i > conv_tol]

        '''
        preconditioning step
        '''
        X_new, Y_new = math_helper.TDDFT_diag_preconditioner(R_x=R_x[index,:],
                                                            R_y=R_y[index,:],
                                                            omega=omega[index],
                                                            hdiag=hdiag)     
                                                   
        '''
        GS and symmetric orthonormalization
        '''
        size_old = size_new
        fill_holder_cost_start = time.time()
        V_holder, W_holder, size_new = fill_holder(V_holder=V_holder,
                                                    W_holder=W_holder,
                                                    X_new=X_new,
                                                    Y_new=Y_new,
                                                    m=size_old,
                                                    double=False)
        fill_holder_step_cost = time.time() - fill_holder_cost_start
        fill_holder_cost += fill_holder_step_cost

        if size_new == size_old:
            print('All new guesses kicked out!!!!!!!')
            break

    davidson_cost = time.time() - davidson_start

    if ii == (max_iter -1) and max_norm >= conv_tol:
        print('=== TDDFT eigen solver not converged due to max iteration mimit ===')
        print('max residual norms', np.max(r_norms))

    print(f'Finished in {ii+1:d} steps, {davidson_cost:.2f} seconds')
    print(f'final subspace = {sub_A.shape[0]}', )
    print(f'max_norm = {max_norm:.2e}')
    for enrty in ['MVcost','fill_holder_cost','subgencost','subcost','full_cost']:
        cost = locals()[enrty]
        print(f'{enrty:<10} {cost:<5.4f}s {cost/davidson_cost:<5.2%}')

    print('======= TDDFT Eigen Solver Done =======' )

    return omega, X_full, Y_full

