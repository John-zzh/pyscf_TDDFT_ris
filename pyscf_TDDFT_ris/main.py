from pyscf_TDDFT_ris import TDDFT_ris, readMO, parameter
import argparse

def str2bool(str):
    if str.lower() in ('yes', 'true', 't', 'y', '1'):
        return True
    elif str.lower() in ('no', 'false', 'f', 'n', '0'):
        return False
    else:
        raise argparse.ArgumentTypeError('Unsupported value encountered.')

def gen_args():
    parser = argparse.ArgumentParser(description='Davidson')
    parser.add_argument('-f',    '--fchfile',     type=str,   default='*.fch',   help='.fch filename (molecule.fch)')
    parser.add_argument('-func', '--functional',  type=str,   default=None,      help='functional filename (pbe0)')
    parser.add_argument('-ax',   '--a_x',         type=float, default=None,      help='HF component in the hybrid functional')
    parser.add_argument('-w',    '--omega',       type=float, default=None,      help='screening factor in the range-separated functional')
    parser.add_argument('-al',   '--alpha',       type=float, default=None,      help='alpha in the range-separated functional')
    parser.add_argument('-be',   '--beta',        type=float, default=None,      help='beta in the range-separated functional')
    
    parser.add_argument('-th',   '--theta',       type=int,   default=0.2,       help='exponent = theta/R^2, optimal theta = 0.2')
    parser.add_argument('-p',    '--add_p',       type=bool,  default=False,     help='add an extra p function to the auxilibary basis')

    parser.add_argument('-tda',  '--TDA',         type=bool,   default=False,     help='peform TDA calculation instead of TDDFT') 
    parser.add_argument('-n',    '--nroots',      type=int,   default=20,        help='the number of states you want to solve')
    parser.add_argument('-t',    '--conv_tol',    type=bool,  default=1e-5,      help='the convengence tolerance in the Davidson diagonalization')
    parser.add_argument('-i',    '--max_iter',    type=int,   default=20,        help='the number of iterations in the Davidson diagonalization')

    args = parser.parse_args()


    if args.functional == None:
        if args.a_x == None and args.omega == None and args.alpha == None and args.beta == None:
            raise ValueError('Please specify the functional name or the functional parameters')
        else:
            if a_x:
                args.a_x = a_x
                print("hybrid functional")
                print(f"manually input HF component ax = {a_x}")

            elif omega and alpha and beta:
                args.a_x = 1
                args.omega = omega
                args.alpha = alpha
                args.beta = beta
                print("range-separated hybrid functional")
                print(f"manually input ω = {args.omega}, screening factor")
                print(f"manually input α = {args.alpha}, fixed HF exchange contribution")
                print(f"manually input β = {args.beta}, variable part")

            else:
                raise ValueError('missing parameters for range-separated functional, please input (w, al, be)')

    elif args.functional != None:
        functional = args.functional.lower()
        args.functional = functional
        print('Loading defult functional paramters from parameter.py.')

        if functional in parameter.rsh_func.keys():
            '''
            RSH functional, need omega, alpha, beta
            '''
            print('use range-separated hybrid functional')
            omega, alpha, beta = parameter.rsh_func[functional]
            args.a_x = 1
            args.omega = omega
            args.alpha = alpha
            args.beta = beta

        elif functional in parameter.hbd_func.keys():
            print('use hybrid functional')
            args.a_x = parameter.hbd_func[functional]

        else:
            raise ValueError(f"I do not have paramters for functional {mf.xc} yet, please either manually input HF component a_x or add parameters in the parameter.py file.")

    return args

args = gen_args()


if __name__ == '__main__':
    

    '''
    if mf object already has a functional name, 
    then do not need to specify the a_x or (omega, alpha, beta), 
    as they will be read from the build-in library

    if manually specify the a_x or (omega, alpha, beta), then you dont have to input the functional name 
    '''
    mf = readMO.get_mf_from_fch(fch_file = args.fchfile, 
                                functional = args.functional)
    
    td = TDDFT_ris.TDDFT_ris(mf=mf, 
                            theta=args.theta,
                            add_p=args.add_p, 
                            a_x=args.a_x,
                            omega=args.omega,
                            alpha=args.alpha,
                            beta=args.beta,
                            conv_tol=args.conv_tol,
                            nroots=args.nroots, 
                            max_iter=args.max_iter)


    if args.TDA == True:
        energies, X, oscillator_strength = td.kernel_TDA()
    else:
        energies, X, Y, oscillator_strength = td.kernel_TDDFT()
