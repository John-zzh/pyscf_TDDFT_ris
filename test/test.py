import numpy as np
from pyscf import gto,  dft
from pyscf_TDDFT_ris import TDDFT_ris_Ktrunc as TDDFT_ris
# from pyscf_TDDFT_ris import TDDFT_ris as TDDFT_ris
np.set_printoptions(linewidth=250, threshold=np.inf, precision=15 )

def gen_mf(RKS, func, charge=0, spin=0):
    mol = gto.Mole()
    mol.verbose = 3
    mol.atom = '''
    C         -4.89126        3.29770        0.00029
    H         -5.28213        3.05494       -1.01161
    O         -3.49307        3.28429       -0.00328
    H         -5.28213        2.58374        0.75736
    H         -5.23998        4.31540        0.27138
    H         -3.22959        2.35981       -0.24953
    '''
    mol.basis = 'def2-SVP'
    mol.build()

    if RKS:
        mf = dft.RKS(mol)
    else:
        mf = dft.UKS(mol) 

    # mf = mf.density_fit()
    mf.conv_tol = 1e-11
    mf.grids.level = 3
    mf.xc = func
    mol.charge = charge
    mol.spin = spin
    mf.kernel()
    return mf

def print_list(list):
    print('[', end='')
    for i in list:
        print(i, end=', ')
    print(']')
    print()

def diff(a,b, threshold=1e-5):
    a = np.asarray(a)
    b = np.asarray(b)
    diff = np.linalg.norm(a-b)
    print(diff, diff<threshold)
    if diff > 1e-5:
        print('wrong results generated')
    
res={}
res['RKS_pbe_TDA_ene']=[6.391337036766483, 7.727736771208482, 8.340518691575443, 8.808272662133794, 9.030071732629553, 9.649731438960863, 10.112653089947635, 10.6661241433674, 10.82618709448345, 10.94062217252083]
res['RKS_pbe_TDA_spc']=[0.002250080569749848, 0.024860626796076726, 0.005520041744598628, 0.02167853938250973, 0.050052098699082365, 0.09100119713743063, 0.0004638288497904391, 0.03490304377710013, 0.01810883404564871, 0.11177763514018192]

res['RKS_pbe_TDDFT_ene']=[6.388712243217593, 7.722596952015948, 8.317118082941937, 8.800279867352872, 9.003292541867337, 9.62323098434012, 10.111323800675956, 10.652787546396684, 10.816234689685365, 10.898194552652988] 
res['RKS_pbe_TDDFT_spc']=[0.00177004727430616, 0.022527165325209306, 0.004680284790039522, 0.018707539453850203, 0.04597997900080101, 0.08023269555633901, 0.0003600754903553672, 0.03183952605457621, 0.017507151309186182, 0.09933956429657763]

res['RKS_pbe0_TDA_ene']=[7.097018627234927, 8.794054508573467, 9.098155376308503, 9.847267859018668, 10.072842892642461, 10.50075087912975, 10.77129403823725, 11.58517494579172, 11.645408335392746, 11.860285837966382]
res['RKS_pbe0_TDA_spc']=[0.0016598202890611687, 0.05060318093802627, 0.0060112846847340164, 0.023492325015694607, 0.056696587735609136, 0.15401547393868073, 0.006235152632582355, 0.0727913082653769, 0.024902926330408956, 0.14788532792421868]

res['RKS_pbe0_TDDFT_ene']=[7.093882111456887, 8.786377038782907, 9.07683721963022, 9.840151658142196, 10.045276834286607, 10.474034332889879, 10.765072347701429, 11.565624341839296, 11.641141860600381, 11.805778194549802]
res['RKS_pbe0_TDDFT_spc']=[0.0012757043082446849, 0.0466835106024071, 0.005276813715839302, 0.020619896132278288, 0.055055943190518646, 0.13645552747311857, 0.0054222403037118404, 0.06613229366726169, 0.022033548843186326, 0.13632967127157392]

res['RKS_wb97x_TDA_ene']=[7.385900261524653, 9.504612486068254, 9.51556248478982, 10.464683242620593, 10.761565519284856, 11.042870164207473, 11.150448550790642, 12.177903133600742, 12.220045646635777, 12.454234955177991]
res['RKS_wb97x_TDA_spc']=[0.0009369273379025906, 0.07372964112083992, 0.004116010850676041, 0.019652959484221046, 0.07221394818126496, 0.18958120157396943, 0.01583996735076162, 0.011732581247642909, 0.10935768241531149, 0.2124429524988815]

res['RKS_wb97x_TDDFT_ene'] = [7.382996952815185, 9.493111644704317, 9.496350313956572, 10.458818721065613, 10.729397345512368, 11.021744554071587, 11.140692725490045, 12.172799158694682, 12.196269740796243, 12.395434222684072]
res['RKS_wb97x_TDDFT_spc'] = [0.0006916521697295814, 0.0036381545119441513, 0.06895820694535056, 0.017501687919707763, 0.0739242834483152, 0.16733215922458056, 0.013774103374428385, 0.009486234006989303, 0.098563678187557, 0.19548415522044307]


res['UKS_pbe_TDA_ene']=[1.483963778598052, 4.099731343007575, 6.042077048481587, 6.691299585099439, 9.801681665134955, 9.946248508827065, 9.992224781839704, 10.339323415696011, 12.081216043877998, 12.147808834323701]
res['UKS_pbe_TDA_spc']=[0.0002770099154354904, 0.00048683852519745103, 0.12876672685788404, 0.013795669440839278, 0.0001636862648334754, 0.003622434380995477, 0.0003570338047231752, 0.027945731744866215, 0.0013766053598939481, 0.025923221522235463]

res['UKS_pbe_TDDFT_ene']= [1.4677594745685272, 4.09913049595788, 5.538681755105241, 6.6589560813707465, 9.800357825462727, 9.932134474004052, 9.992095480814076, 10.311037517482896, 12.080908467668143, 12.143126046039484]
res['UKS_pbe_TDDFT_spc']=[7.180743655444574e-05, 0.000591781174062631, 0.07187972790996885, 0.00960551649818776, 0.00023085639787773641, 0.0038903329401893923, 0.00034446092120029843, 0.022349365438394473, 0.0015551691427012492, 0.025233871951721974]

res['UKS_pbe0_TDA_ene']=[1.8087589919782452, 4.405998297362837, 5.72302982858523, 6.92662003448187, 10.481419712347002, 10.63128573558761, 10.782950917081239, 11.23327456870174, 12.668737025010973, 12.681847067062224]
res['UKS_pbe0_TDA_spc']=[0.0001657568385854985, 0.001076181852041418, 0.12111409902871455, 0.011301537524469637, 8.610279495946372e-05, 0.0030344766658605064, 0.0014924564272755585, 0.027662168245754536, 0.009330311094434413, 7.336079775827479e-06]

res['UKS_pbe0_TDDFT_ene']=[1.7942558935943167, 4.3977767134727275, 5.400199237079016, 6.896833487311187, 10.47866239615581, 10.621547843026367, 10.781601271341614, 11.20807454635799, 12.662671164284594, 12.680077165587111]
res['UKS_pbe0_TDDFT_spc']=[3.349443421396765e-05, 0.0010872607187901665, 0.0770981778826746, 0.007894982491473182, 0.0001509349609445147, 0.0032931485229418397, 0.0014241487059804376, 0.022564345593069635, 0.009241371630422714, 8.096157694307589e-06]

res['UKS_wb97x_TDA_ene']=[1.7801959249879324, 4.366274710265926, 5.630072600091703, 6.863280584494665, 10.960569116608673, 11.086243641692363, 11.141070826270319, 11.44658612100384, 13.04900918223636, 13.052876348291248]
res['UKS_wb97x_TDA_spc']=[0.0001310268269702344, 0.0012363348261031622, 0.1202749786520348, 0.010936137627666015, 1.7928572229891145e-05, 0.0007151939265619905, 0.0018611629224512599, 0.032169117965574145, 0.006232586599722929, 0.0010126781058545358]

res['UKS_wb97x_TDDFT_ene'] = [1.763067543956076, 4.353080374822271, 5.3338409882235025, 6.832369369409955, 10.95577150841456, 11.084260676422906, 11.130693620714675, 11.418990183153687, 13.03973729659376, 13.049827577160876]
res['UKS_wb97x_TDDFT_spc'] = [2.4956123757873378e-05, 0.0012283555854364617, 0.07938975994932368, 0.007720883972830806, 6.845912316793527e-05, 0.0007065956504526979, 0.002058534987298801, 0.026340830374532796, 0.006611670493927908, 0.001019094231403822]


def main():

    nroots = 20
    for KS in ['RKS', 'UKS']:
    # for KS in ['RKS']:
        for func in ['pbe', 'pbe0', 'wb97x']:
        # for func in ['pbe']:
            for calc in ['TDA', 'TDDFT']:
            # for calc in ['TDDFT']:
                name = KS + '_' + func + '_' + calc
                print('======================================= {} {} {}-ris ======================================='.format(KS, func, calc))
                mf = gen_mf(RKS=True if KS=='RKS' else False, func=func, charge=0 if KS=='RKS' else 1, spin=0 if KS=='RKS' else 1)
                td = TDDFT_ris.TDDFT_ris(mf, nroots=nroots, spectra=False)
                if calc == 'TDA':
                    energies, X, oscillator_strength = td.kernel_TDA()
                elif  calc == 'TDDFT':
                    energies, X, Y, oscillator_strength = td.kernel_TDDFT()
                # print_list(energies)
                # print_list(oscillator_strength)
                print(name)
                compare_nroors = min(nroots, 10)
                diff(energies[:compare_nroors], res[name+"_ene"][:compare_nroors])
                diff(oscillator_strength[:compare_nroors], res[name+"_spc"][:compare_nroors])
   
def main2():
    nroots = 20
    # for KS in ['RKS', 'UKS']:
    for KS in ['RKS']:
        # for func in ['pbe', 'pbe0', 'wb97x']:
        for func in ['wb97x']:
            # for calc in ['TDA', 'TDDFT']:
            for calc in ['TDA','TDDFT']:
                name = KS + '_' + func + '_' + calc
                print('======================================= {} {} {}-ris ======================================='.format(KS, func, calc))
                mf = gen_mf(RKS=True if KS=='RKS' else False, func=func, charge=0 if KS=='RKS' else 1, spin=0 if KS=='RKS' else 1)
                td = TDDFT_ris.TDDFT_ris(mf, nroots=nroots, spectra=True, Ktrunc=0, J_fit='sp', K_fit='s')
                if calc == 'TDA':
                    energies, X, oscillator_strength = td.kernel_TDA()
                    # print(energies)
                elif  calc == 'TDDFT':
                    energies, X, Y, oscillator_strength = td.kernel_TDDFT()
                # print_list(energies)
                # print_list(oscillator_strength)
                print(name)
                compare_nroors = min(nroots, 10)
                diff(energies[:compare_nroors], res[name+"_ene"][:compare_nroors])
                diff(oscillator_strength[:compare_nroors], res[name+"_spc"][:compare_nroors], threshold=0.01)    


if __name__ == '__main__':
    main2()
