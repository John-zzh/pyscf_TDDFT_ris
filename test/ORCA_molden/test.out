This job can use: 8CPUs
converged SCF energy = -114.283776041434
n_occ = 8
n_vir = 30
asigning auxiliary basis set, add p function = False
U = 0.2
asigning auxiliary basis set, add p function = False
U = 0.2
2c2e and 3c2e for (ia|jb)
eri2c_cl.shape (4, 4)
eri3c_cl.shape (38, 38, 4)
step  1 max_norm = 0.19189737842722424
step  2 max_norm = 0.006747503144944444
step  3 max_norm = 0.0008507430371836815
step  4 max_norm = 6.11378312642584e-05
step  5 max_norm = 4.259123071985376e-06
step  6 max_norm = 1.325941356190859e-07
step  7 max_norm = 4.272860055424579e-09
step  8 max_norm = 1.2260015133189845e-10
step  9 max_norm = 4.059650853707374e-11
TDDFT eigen solver Guess Done
Finished in 9 steps, 0.12 seconds
final subspace 149
max_norm = 4.06e-11
MVcost     0.0194s 15.74%
GScost     0.0365s 29.54%
subgencost 0.0124s 10.06%
subcost    0.0370s 29.96%
full_cost  0.0057s 4.59%
spectra written to ***REMOVED*** 

[ 3.82830377  8.45004826  8.7483101   9.43236068 10.04881749 11.03618182
 11.11825733 11.83038896 13.32050668 13.48012317 13.82619431 14.24302986
 14.59352031 14.90991819 16.19884469 16.43967788 17.22090509 17.96167157
 18.35158112 18.54198394]
==================
n_occ = 8
n_vir = 30
asigning auxiliary basis set, add p function = False
U = 0.2
asigning auxiliary basis set, add p function = False
U = 0.2
2c2e and 3c2e for (ia|jb)
eri2c_cl.shape (4, 4)
eri3c_cl.shape (38, 38, 4)
step  1 max_norm = 0.2891157940336989
step  2 max_norm = 0.030127235308202226
step  3 max_norm = 0.004581532117651838
step  4 max_norm = 0.0004284584372311571
step  5 max_norm = 7.210151977624313e-05
step  6 max_norm = 2.7654861435029803e-06
step  7 max_norm = 5.623808325148691e-08
step  8 max_norm = 1.3062836566160274e-09
step  9 max_norm = 5.000381517141588e-11
TDDFT eigen solver Guess Done
Finished in 9 steps, 12.22 seconds
final subspace 156
max_norm = 5.00e-11
MVcost     12.0989s 99.04%
GScost     0.0325s 0.27%
subgencost 0.0227s 0.19%
subcost    0.0424s 0.35%
full_cost  0.0063s 0.05%
spectra written to TDDFT-abinitio_UV_spectra.txt 

[ 4.02201852  8.47259058  9.19685489  9.68880836 10.26842207 11.16723882
 11.23655156 11.9400954  13.52538676 13.53495054 14.42283844 14.78185573
 15.05723447 15.63410156 16.42347194 16.70636113 17.61015659 18.32463158
 18.65912522 18.70155506]
==================
